### DISCUSSION TITLE
Support for PyQtGraph Plotting Capability

### DISCUSSION THREAD

**Discussion by j9ac9k** [networkx/networkx#4879] [Ideas]
Greetings from PyQtGraph land.  

I'm continuing the [discussion started on the mail list](https://groups.google.com/d/msgid/networkx-discuss/CA%2BXMcTN4kWOxW%2BPLZ5bZZeqEuQyiGR-GkUwPz9Kpf5JwxcUZNw%40mail.gmail.com?utm_medium=email&utm_source=footer) regarding evaluating the work/effort involved to have PyQtGraph be supported as a plotting backend for NetworkX.  

PyQtGraph has some (very) limited capability of plotting graph data presently.  Where PyQtGraph does accelerate in is in areas of interactivity, such as mouse zoom/hover and in the case of 3D, rotate.  PyQtGraph also accelerates in areas involving (rapidly) updating data.  Where PyQtGraph does not accelerate so much is in static plot images, that is an area other libraries (graphviz, matplotlib and so on) will likely meet user needs better than pyqtgraph.

To that end, PyQtGraph's current graph plot capability is somewhat limited, we are overly restrictive about the data-structures that can be passed in to a graph plot to correctly draw nodes/edges and so on.  To that end, I am happy to make changes on PyQtGraph's end to support data structures/patterns common to NetworkX, so at least the libraries can be used together with minimal effort on the users end.  I figure this is likely a good first step.

Lastly, I do have a question, is there some plotting capability that is missing from the current options supported?   I would be happy to try and evaluate if PyQtGraph could meet those respective use-cases.

Thanks for maintaining this library!

**Comment by dschult** [comment]
Thanks for this!

The matplotlib interface bases our node-drawing using a scatterplot tool. That might be very similar to what PyQtGraph is doing.  Edges are currently handled using one of two possible drawing tools -- either a collection of lines (one for each edge), or as a list of FancyArrowPatches (one for each edge). The FancyArrowPatch allows arrowheads, bending curves instead of lines, and adjustments of the arrowheads along the edge so the tip lines up with the edge of whatever shape we used to draw the nodes.
Finally, there is the ability to add text to label the nodes and the edges. Other than many options and features which can be controlled by the users, these are the basic tools we use from Matplotlib. The speed crunch is typically in handled the large number of edges (often on the order of the square of the number of nodes).

Does it make sense to try to co-opt the drawing interface for Matplotlib to make a parallel set of functions for PyQtGraph? Or would it be better to let PyQtGraph construct tools and an interface that slurps up the NetworkX graph uses what it needs to build the drawing?  If we can import node/position data and then a list of 2-tuples of node-pairs which form edges, are your tools ready to draw a shape at each node and then connect the positions for the nodes with curves to denote the edges?

As for missing capabilities on the NetworkX drawing side, we don't have good interactive drawing capabilities. And we have struggled for a long time to get arrows to behave well with various node shapes; to figure out the best way to draw self-loops; and to allow curved edges that provide nice visualizations.

**Reply by j9ac9k** [reply]
Hi @dschult I'll try and answer questions in order; there's a lot of great topics and I don't want to miss anything.

> The matplotlib interface bases our node-drawing using a scatterplot tool. That might be very similar to what PyQtGraph is doing. Edges are currently handled using one of two possible drawing tools -- either a collection of lines (one for each edge), or as a list of FancyArrowPatches (one for each edge). The FancyArrowPatch allows arrowheads, bending curves instead of lines, and adjustments of the arrowheads along the edge so the tip lines up with the edge of whatever shape we used to draw the nodes.

Fancy arrows I suspect are going to be the biggest hang-up on PyQtGraph's end.  While PyQtGraph has no trouble drawing lines/arrows, I'm not sure if it has the capability of drawing curved arrows.  I'll have to investigate this more.

> Finally, there is the ability to add text to label the nodes and the edges. 

PyQtGraph has this ability.  Here is the output of our [CustomGraphItem example](https://github.com/pyqtgraph/pyqtgraph/blob/master/examples/CustomGraphItem.py)

![image](https://user-images.githubusercontent.com/646398/121081220-41058c00-c791-11eb-9285-feb384dd2abb.png)

I could support the embedding of text into the base class for GraphItem plots, we have a TextIem object that stores text and can be anchored to other objects.  The example also does allow for moving nodes through mouse drag events.

![image](https://user-images.githubusercontent.com/646398/121081444-95a90700-c791-11eb-9944-8de692f4e6f8.png)

> The speed crunch is typically in handled the large number of edges (often on the order of the square of the number of nodes).

If you want to get a ballpark on how quickly PyQtGraph is with respect to number of nodes, we have a ScatterPlotSpeedTest benchmark file in our examples, which has some options, including number of points, you can see how the FPS changes when adding more points.   With pyqtgraph installed (currently pyqtgraph is quite broken with PySide 6.1.1, which was released earlier today, but works w/ PySide6 6.1.0) you can run `python -m pyqtgraph.examples` and under the benchmark sections run the ScatterPlotSpeedTest.

Regarding edges, ...I'll have to think how performance will very here.  We have some very fast methods for drawing lines, but if we want to vary the color/thickness by segment, this complicates things a bit (from a performance standpoint).  I'd have to profile more, but I suspect the number of nodes will be the bigger factor in determining performance.

> Does it make sense to try to co-opt the drawing interface for Matplotlib to make a parallel set of functions for PyQtGraph? Or would it be better to let PyQtGraph construct tools and an interface that slurps up the NetworkX graph uses what it needs to build the drawing?

I think having PyQtGraph's standard API be able to take standard data structures that NetworkX uses is a logical first step.

1. It (should) minimize work needed on the NetworkX end
2. Should make it easier to evaluate usability/feasibility
3. Even if not entirely feasible for integration, should still result in an improved experience for pyqtgraph users drawing plots that use NetworkX

> If we can import node/position data and then a list of 2-tuples of node-pairs which form edges, are your tools ready to draw a shape at each node and then connect the positions for the nodes with curves to denote the edges?

Right now, this is what pyqtgraph needs to draw a graph: https://github.com/pyqtgraph/pyqtgraph/blob/master/pyqtgraph/graphicsItems/GraphItem.py#L27-L55

PyQtGraph wants a (N, 2) array for x, y positions in the view-space to place each node, it wants a (M, 2) array where each row contains the indexes of the two nodes that are being connected.  There are other arguments regarding what symbol to draw for each node,  size, colors, and so on.  To more directly answer your question, I think we have more or less everything you need here assuming your 2-tuples of node-pairs reference nodes by index in their position argument.

> As for missing capabilities on the NetworkX drawing side, we don't have good interactive drawing capabilities. And we have struggled for a long time to get arrows to behave well with various node shapes; to figure out the best way to draw self-loops; and to allow curved edges that provide nice visualizations.

As I said earlier PyQtGraph doesn't really have support for curved edges, I can see this being a tough problem to solve; but there is native Qt support for cubic and quadratic Bezier curves which wouldn't be too difficult for us to implement.  Making it look nice in visualizations across the board is, as I'm sure you know, not going to be an easy problem to solve.

**Reply by j9ac9k** [reply]
I would happily create a script you can run/experiment with, is there some kind of test dataset that you like to use?  I'll gladly write up something that would inject the output from NetworkX and render it on the screen for others to examine.

**Reply by dschult** [reply]
Most layout functions we provide allow an argument for the number of dimensions for the layout. But some do not yet have the code to handle 3-D. Many do...   The output of these functions is a dict keyed by node to the node's position (x,y). This should be easy to convert to an Nx2 array of x/y positions. We even have code inside a function that does that conversion but I forget if it is isolated as a useable separate function. Seems reasonable to do so if it doesn't already exist.  The 2-tuple edges I was referring to are indeed pairs of nodes so that should be easy to convert to an array as well.

```python
pos = nx.spring_layout(G)
node_pos_array = np.array(list(pos.values()))

node_index = {node: i for i, node in enumerate(G)}
edge_list_array = np.array(list((node_index[u], node_index[v]) for u,v in G.edges))
```
This code would convert the position data from our layout functions into a node position array for you. And the edge array requires a way to get the node index -- hence the dict `node_index`

As far as a test dataset, I'm not sure what a favorite would be.  Choose one from [our Gallery](https://networkx.org/documentation/stable/auto_examples/index.html#drawing).
Maybe "node colormap" or "circular tree" for smaller networks or "random geometric network" for something larger.

**Comment by rossbar** [comment]
I really like this idea - I've used pyqtgraph in the past (thanks for the awesome project!) for computationally-intensive visualization and was always impressed with a) the quality of the output and b) the performance.

NetworkX's drawing "backends" don't match the criteria of what I typically think of as a "backend", which is more akin to the way matplotlib is organized --- the key IMO is that matplotlib provides the *consistent interface for users* that then is implemented via the different backends (TkAgg, WxAgg, Qt5Agg, etc.). Thus the user creates their visualization in the same way (e.g. pyplot.plot) regardless of implementation, but then can customize the resulting object(s) via the interface offered by the specific backend if so desired.

I could envision a similar type of organization for the `networkx.drawing` package. Basically, networkx defines the user-facing API (e.g. `nx.draw_networkx_nodes`) which can implemented in different "backends" e.g. `matplotlib.pyplot`, `pyqtgraph`, etc. In the long run I think this would be very valuable.

This is all very high-level though and more of a long-term idea. In the short term I think it would be great to graph visualization applications with pyqtgraph. One obvious place to do so would be in the example gallery.

**Reply by j9ac9k** [reply]
Hi @rossbar 

I'm glad to hear that pyqtgraph has worked out for you well in the past.  We (pyqtgraph maintainers) know remarkably little about our user-base; and always love to hear about use-cases, pain points, and so on.  If you have a project that uses pyqtgraph and want to give feedback, or discuss how improvements in some places of the library would positively impact you and your work, feel free to open a new discussion in our repo.

> NetworkX's drawing "backends" don't match the criteria of what I typically think of as a "backend", which is more akin to the way matplotlib is organized --- the key IMO is that matplotlib provides the consistent interface for users that then is implemented via the different backends (TkAgg, WxAgg, Qt5Agg, etc.). Thus the user creates their visualization in the same way (e.g. pyplot.plot) regardless of implementation, but then can customize the resulting object(s) via the interface offered by the specific backend if so desired.

I largely agree with this; but the wording I brought up stems from [pandas plotting "backends"](https://pandas.pydata.org/pandas-docs/dev/user_guide/visualization.html#plotting-backends), which we will likely start work on supporting at some point in the future (although pandas requires we have a lot more plot types than we currently offer, and pandas data, while could benefit from mouse interactivity, I don't see benefitting from rapidly updating data).  

> This is all very high-level though and more of a long-term idea. In the short term I think it would be great to graph visualization applications with pyqtgraph. One obvious place to do so would be in the example gallery.

I'll start working on recreating some of these.

Right now our graph plot interface is really bare-bones, and instead of me speculating about how to better expand and make it more versatile in a vacuum, figured might be better to reach out to you all; and see see what kind of data structures are common in NetworkX, and try and make pyqtgraph work with those a bit more seamlessly.

Question/Confirmation, looks like your graph layout methods have the capability of returning coordinates in 3D space, not just 2D, is this the case for all layouts, or is it more limited.

**Comment by j9ac9k** [comment]
Hey everyone, after a delay due to planning for the SciPy sprints (our first!) I've started working on this as part of our (PR #1721)[https://github.com/pyqtgraph/pyqtgraph/pull/1721]

I have a _super_ basic 3D plot example here (you need the install the library from that PR to replicate)

```python
import pyqtgraph as pg
from pyqtgraph.Qt import QtCore, QtGui
import pyqtgraph.opengl as gl
import numpy as np

import networkx as nx

app = pg.mkQApp("GLGraphItem Example")
w = gl.GLViewWidget()
w.opts["distance"] = 20
w.show()

# networkX degree sequence example
# https://networkx.org/documentation/stable/auto_examples/graph/plot_degree_sequence.html
z = [5, 3, 3, 3, 3, 2, 2, 2, 1, 1, 1]
G = nx.configuration_model(z)  # configuration model
degree_sequence = [d for n, d in G.degree()]  # degree sequence
hist = {}
for d in degree_sequence:
    if d in hist:
        hist[d] += 1
    else:
        hist[d] = 1

edges = np.array(list(G.edges()))
positions = np.array(list(nx.drawing.layout.spring_layout(G, dim=3).values()))
edgeColor=pg.glColor("w")

gi = gl.GLGraphItem(
    edges=edges,
    nodePositions=positions,
    edgeWidth=1.,
    nodeSize=10.
)

w.addItem(gi)

if __name__ == "__main__":
    pg.exec()
```

https://user-images.githubusercontent.com/646398/126873533-8494c0be-796a-4de0-9ebb-1ee21e429024.mov

Things are fairly straight forward here.  PyQtGraph by default wants numpy arrays representing node positions and edge connectivity (because we feed those arrays straight to openGL calls); but that's straight forward to convert from networkX's data structure. 

Next up, I need to extract styling information, and map that accordingly.

I'm debating if I want to introduce code within the pyqtgraph to take networkX graph objects and the output of `networkx.drawing.layout` and go from there; but I can punt on that right now, next up, styling and potentially annotation information is next.

If there is a particular plot in the gallery you think would be interesting to have in some kind of 3D/interactive plot, let me know and I'll do that next.
